### ISSUE TITLE
Defer scroll position update for async data loads

### ISSUE THREAD

**Issue by AndrewIngram** [taion/scroll-behavior#21]
It's hard to identify where in the chain this is breaking. I'm using react-router and react-router-relay.

### What happens

* I click a link
* The address bar updates and the browser scrolls to the top (hard to say what order these two things happen in)
* Data loads
* The new component renders

### What I expect (to mimic non client-side routing)

* I click a link
* Data loads
* The new component renders, address bar is updated and browser scrolls to the top

### Alternative

* I click a link
* Address bar updates
* Data loads
* New component renders, browser scrolls to the top

TLDR; The scroll-to-top happens immediately and is applied to the current route, rather than the new route, which is confusing my users.

**Comment by taion** [comment]
Known bug, unfixable with current API. We'll get something new out once the updated router API lands.

**Comment by hanjukim** [comment]
In my case, this bug causes scroll position does not get restored. I'm using standard scroll behavior.

# What Happens
* Current path is '/' and scroll position is [0, 15000]
* History push to '/new'
* Address updates, scroll position of '/new' is not stored. Scroll to [0, 0] and store position.
* Page renders
* History goBack to '/'
* Address updates, scroll position of '/' is stored at [0, 15000]. Calling window.scrollTo(0, 15000); but '/new' page has shorter window height. Notting happens.

# What I Expect
* History push
* Store scroll position
* Address update
* Page renders
* Scroll restores

# Alternative
* Wrap window.scrollTo into setTimeout to scroll after actual page change.

**Comment by blackbing** [comment]
Finally I write another simple library to avoid this issue. hope it could be helpful.  https://github.com/blackbing/use-scroll-behavior

**Comment by jackyon** [comment]
+1, is there any solution for this? it should not be scroll to top immediately until the new page finish loading.

**Comment by blackbing** [comment]
@jackyon try [use-scroll-behavior](https://github.com/blackbing/use-scroll-behavior)?

**Comment by taion** [comment]
This cannot be fixed with the current API – it's exactly why the library needs to tie into the router rather than the history.

**Comment by jackyon** [comment]
hi @taion, is there any schedule for this?

**Comment by taion** [comment]
My current thought is to add a callback-based async option to the `shouldScrollUpdate` callback, which would let us handle this sort of thing.

**Comment by michalmikolajczyk** [comment]
I bumped into this too now. Is this being addressed?

**Comment by taion** [comment]
Not yet.

**Comment by mmahalwy** [comment]
@taion is there ETA? how could we help? This is a major blocker.

**Comment by koichik** [comment]
@taion How about this idea?

```diff
diff --git a/modules/utils/createUseScroll.js b/modules/utils/createUseScroll.js
index ecc9998..2c4044d 100644
--- a/modules/utils/createUseScroll.js
+++ b/modules/utils/createUseScroll.js
@@ -7,12 +7,14 @@ import requestAnimationFrame from 'dom-helpers/util/requestAnimationFrame'
 // Try at most this many times to scroll, to avoid getting stuck.
 const MAX_SCROLL_ATTEMPTS = 2

+const defaultScrollTo = (x, y) => window.scrollTo(x, y)
+
 export default function createUseScroll(
   getScrollPosition, start, stop, updateLocation
 ) {
   return function (createHistory) {
     return function (options = {}) {
-      const { shouldUpdateScroll, ...historyOptions } = options
+      const { shouldUpdateScroll, scrollTo, ...historyOptions } = { scrollTo: defaultScrollTo, ...options}

       const history = createHistory(historyOptions)

@@ -97,7 +99,7 @@ export default function createUseScroll(
         }

         const [ x, y ] = scrollTarget
-        window.scrollTo(x, y)
+        scrollTo(x, y)

         ++numScrollAttempts
```
It's a bit adhoc, but we can save scroll position and restore it after data loads.

**Comment by taion** [comment]
No, the way this will work is that `shouldUpdateScroll` will have an optional async callback, to defer firing until the router renders the new scene.

**Comment by michalmikolajczyk** [comment]
I managed to solve this with a workaround.

I initially was getting the problem, while navigating after clicking on a `Link` component:

```
<Link to={somePathVar}>Click to navigate</Link>
```

then I noticed, then when using pushState, the problem does not occur. So I changed the `Link` component to good ol' `a`:

```
<a onClick={this.onLinkClick(somePathVar} href={somePathVar}>Click to navigate</a>
```

**I believe that the order of onClick and href is quite important here**

the onLinkClick() class method looks like this:

```
  onLinkClick(route) {
    return (ev) => {
      ev.preventDefault();
      ev.stopPropagation();
      this.props.pushState(route);
    };
  }
```

some additional info: 
* I use `pushState`, which actually is `routeActions.push` from `react-router-redux`, to be dispatched
* I obviously have `onLinkClick` bound to `this` in the constructor:
    ```
    constructor(props) {
      this.onLinkClick = this.onLinkClick.bind(this);
    }
    ```

Hope this is of some help!
